syntax = "proto3";

package dht;

message NodeId {
  bytes id = 1;
  string ip = 2;
  int32 port = 3;
}

message GetValueRequest {
  bytes key = 1;
}
message GetValueResponse {
  string value = 1;
  bool found = 2;
}

message PingRequest {
  NodeId sender = 1;
}

message PingResponse {
  string message = 1;
}

message StoreRequest {
  NodeId sender = 1;
  bytes key = 2;
  bytes value = 3;
}

message StoreResponse {
  string status = 1;
}

message FindNodeRequest {
  NodeId sender = 1;
  bytes target_id = 2;
}

message FindNodeResponse {
  repeated NodeId closest_nodes = 1;
}

message ReplicateChunkRequest {
  NodeId sender = 1;
  bytes key = 2;
  bytes value = 3;
}

message ReplicateChunkResponse {
  string status = 1;
}

message AddPeerRequest {
  NodeId sender = 1;
}

message AddPeerResponse {
  string status = 1;
}

service DHTNode {
  rpc Ping(PingRequest) returns (PingResponse);
  rpc Store(StoreRequest) returns (StoreResponse);
  rpc FindNode(FindNodeRequest) returns (FindNodeResponse);
  rpc ReplicateChunk(ReplicateChunkRequest) returns (ReplicateChunkResponse);
  rpc AddPeer(AddPeerRequest) returns (AddPeerResponse);
  rpc GetValue(GetValueRequest) returns (GetValueResponse);
}